design/userdefined-rootdefinehints.md

```diff
@@ -0,0 +1,101 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# userdefined-rootdevicehints
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [userdefined-rootdevicehints](#userdefined-rootdevicehints)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: Tue Jun 25 14:54:44 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+This document explains how rootdevicehints could be given as user defined parameters as part BaremetalHostSpec. The idea is that user would have a possibilty to defines selective constrains for root device selection.
+
+## Motivation
+
+We need to be able select root device, when there is multiple devices available so that device is not chosen randomly
+
+### Goals
+
+1. Decide root device hints are needed be supported.
+2. Agree on the way haw information is propagated to ironic through BMO
+
+
+### Non-Goals
+
+The complete set of root device hints as input is not needed only those seen important.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- BaremetalHostSpec need to be updated to support list of following root device hints:
+* vendor
+* size
+* wwn
```

Comment:
Wouldn't vendor also lead to non-deterministic behavior?

Is the size used to select location or to set the size of the filesystem? If the former, that also seems likely to allow non-deterministic behavior.

Reply:
If user would like to select any device from dell that match side of the file system, I think that would be OK for our needs, not completely deterministic, but selective enough

Reply:
As a sample in my case vendor was important. Because i had an SD card on the middle and a hard disk . I could set a hint with "vendor:ATA", so it skipped my SD card

Reply:
We have to keep in mind that 90% of the time, or more, no human being is going to be touching the BareMetalHost. Most of the time, the machine API uses it. 

So, if we're going to add fields for hints like this, do they also need to go into the MachineSet's template for Machines so that the machine-api code can copy them into the host during provisioning? Or do we expect the user to pre-configure this hint somehow before allowing the machine API code to touch the host?

If we place "smart" logic like not deploying to removable SD cards into the thing invoking the BareMetalHost API, then we can keep the host API very declarative. For example, the machine API code could look at the disks discovered during inspection and pick one, then provide the hint to set on the host during provisioning.

Reply:
I think we can update CAPBM logic if needed to propagate hints in Machine data, but for BMO could the hints option exposure look as proposed in this doc or?

Reply:
> I think we can update CAPBM logic if needed to propagate hints in Machine data, but for BMO could the hints option exposure look as proposed in this doc or?

I'm honestly not sure. I need to understand how they would be used end-to-end.

If the CAPBM code is able to look at introspection data for a host, we would only need 1 hint type that allowed CAPBM to choose one of the available disks. And that doesn't even need to be one that Ironic supports, since the baremetal-operator could translate a simple index into the list of disks on a host to the relevant hint for the call to Ironic.

Reply:
So did I got it right, that after inspection, BMH object status field should contain a list of disk devices with details so that CAPBM or other controller could should one of those by just giving index?

Reply:
> So did I got it right, that after inspection, BMH object status field should contain a list of disk devices with details so that CAPBM or other controller could should one of those by just giving index?

Yes, that's correct. The status fields for the host include a bunch of hardware details, including the disk info. If some property of a disk isn't present but is available through inspection, we can update the CRD to include it.

Reply:
While trying to patch types, I just started to wonder that why do we need a separate index table, if we already have storage disk table and each has unique name, which could be given in as hint in provisioning phase?

Reply:
> > So did I got it right, that after inspection, BMH object status field should contain a list of disk devices with details so that CAPBM or other controller could should one of those by just giving index?
> 
> Yes, that's correct. The status fields for the host include a bunch of hardware details, including the disk info. If some property of a disk isn't present but is available through inspection, we can update the CRD to include it.

While trying to patch types, I just started to wonder that why do we need a separate index table, if we already have storage disk table and each has unique name, which could be given in as hint in provisioning phase

Reply:
> > > So did I got it right, that after inspection, BMH object status field should contain a list of disk devices with details so that CAPBM or other controller could should one of those by just giving index?
> > 
> > 
> > Yes, that's correct. The status fields for the host include a bunch of hardware details, including the disk info. If some property of a disk isn't present but is available through inspection, we can update the CRD to include it.
> 
> While trying to patch types, I just started to wonder that why do we need a separate index table, if we already have storage disk table and each disk has a unique name, which could be given in as hint in provisioning phase? Our use case is such that we would need to be able to pre-configure host with certain root device hint, before Machine API takes over. i.e. root device selection should be able to do even before inspection phase. Would that be OK that we would have both pre and post inspection root device selection opportunity?

We aren't going to always be able to rely on the user to manually configure everything. In fact, we will have something far more scalable if we minimize the amount of manual configuration of individuals hosts that we require. We have work in progress to automatically classify hosts based on inspection data. We also have work in progress to automatically scale a MachineSet when the new controller finds a host that would match the selection criteria for that MachineSet. When that work is done, adding a host CR will trigger provisioning a host through the machine set automatically. We therefore need to consider the MachineSet as the primary configuration interface that users will interact with when they specify how hosts should be treated. They will tell us "when you see a host that looks like X, add a Machine to set Y and use that Machine to provision the host".

I proposed that we use the index of the device on the host that should receive the root image. The MachineSet would have a field like rootDiskIndex that might have a value like 0 or 1. That single integer would be copied to the BareMetalHost and the baremetal-operator would look at the known disks, find the one with the given index, and extract the relevant hint information to pass to Ironic (HCTL, serial number, whatever we choose).

Maybe using the index is too abstract, though. Maybe we do want the device name or HCTL or something like that. However, at the MachineSet layer we don't want to be using serial numbers or WWNs, because those are specific to individual devices and can't be reused across hosts.

I'm definitely open to other suggestions, but whatever we choose has to work as a setting on a MachineSet, passed to a Machine, and then to a BareMetalHost.

Reply:
> > > > So did I got it right, that after inspection, BMH object status field should contain a list of disk devices with details so that CAPBM or other controller could should one of those by just giving index?
> > > 
> > > 
> > > Yes, that's correct. The status fields for the host include a bunch of hardware details, including the disk info. If some property of a disk isn't present but is available through inspection, we can update the CRD to include it.
> > 
> > 
> > While trying to patch types, I just started to wonder that why do we need a separate index table, if we already have storage disk table and each disk has a unique name, which could be given in as hint in provisioning phase? Our use case is such that we would need to be able to pre-configure host with certain root device hint, before Machine API takes over. i.e. root device selection should be able to do even before inspection phase. Would that be OK that we would have both pre and post inspection root device selection opportunity?
> 
> We aren't going to always be able to rely on the user to manually configure everything. In fact, we will have something far more scalable if we minimize the amount of manual configuration of individuals hosts that we require. We have work in progress to automatically classify hosts based on inspection data. We also have work in progress to automatically scale a MachineSet when the new controller finds a host that would match the selection criteria for that MachineSet. When that work is done, adding a host CR will trigger provisioning a host through the machine set automatically. We therefore need to consider the MachineSet as the primary configuration interface that users will interact with when they specify how hosts should be treated. They will tell us "when you see a host that looks like X, add a Machine to set Y and use that Machine to provision the host".
> 
> I proposed that we use the index of the device on the host that should receive the root image. The MachineSet would have a field like rootDiskIndex that might have a value like 0 or 1. That single integer would be copied to the BareMetalHost and the baremetal-operator would look at the known disks, find the one with the given index, and extract the relevant hint information to pass to Ironic (HCTL, serial number, whatever we choose).
> 
> Maybe using the index is too abstract, though. Maybe we do want the device name or HCTL or something like that. However, at the MachineSet layer we don't want to be using serial numbers or WWNs, because those are specific to individual devices and can't be reused across hosts.
> 
> I'm definitely open to other suggestions, but whatever we choose has to work as a setting on a MachineSet, passed to a Machine, and then to a BareMetalHost.

Do you mean that in Machineset the device index/Name/HCTL number would be the same in all  hosts that has label matching the MachineSet host requirements?

Reply:
> Do you mean that in Machineset the device index/Name/HCTL number would be the same in all hosts that has label matching the MachineSet host requirements?

Yes, that's the idea.

What do you think?

Reply:
> > > > > So did I got it right, that after inspection, BMH object status field should contain a list of disk devices with details so that CAPBM or other controller could should one of those by just giving index?
> > > > 
> > > > 
> > > > Yes, that's correct. The status fields for the host include a bunch of hardware details, including the disk info. If some property of a disk isn't present but is available through inspection, we can update the CRD to include it.
> > > 
> > > 
> > > While trying to patch types, I just started to wonder that why do we need a separate index table, if we already have storage disk table and each disk has a unique name, which could be given in as hint in provisioning phase? Our use case is such that we would need to be able to pre-configure host with certain root device hint, before Machine API takes over. i.e. root device selection should be able to do even before inspection phase. Would that be OK that we would have both pre and post inspection root device selection opportunity?
> > 
> > 
> > We aren't going to always be able to rely on the user to manually configure everything. In fact, we will have something far more scalable if we minimize the amount of manual configuration of individuals hosts that we require. We have work in progress to automatically classify hosts based on inspection data. We also have work in progress to automatically scale a MachineSet when the new controller finds a host that would match the selection criteria for that MachineSet. When that work is done, adding a host CR will trigger provisioning a host through the machine set automatically. We therefore need to consider the MachineSet as the primary configuration interface that users will interact with when they specify how hosts should be treated. They will tell us "when you see a host that looks like X, add a Machine to set Y and use that Machine to provision the host".
> > I proposed that we use the index of the device on the host that should receive the root image. The MachineSet would have a field like rootDiskIndex that might have a value like 0 or 1. That single integer would be copied to the BareMetalHost and the baremetal-operator would look at the known disks, find the one with the given index, and extract the relevant hint information to pass to Ironic (HCTL, serial number, whatever we choose).
> > Maybe using the index is too abstract, though. Maybe we do want the device name or HCTL or something like that. However, at the MachineSet layer we don't want to be using serial numbers or WWNs, because those are specific to individual devices and can't be reused across hosts.
> > I'm definitely open to other suggestions, but whatever we choose has to work as a setting on a MachineSet, passed to a Machine, and then to a BareMetalHost.
> 
> Do you mean that in Machineset the device index/Name/HCTL number would be the same in all hosts that has label matching the MachineSet host requirements? In other words, after inspections, storage device list would be identical for all hosts having the label that match Machineset requirements

If labeling ensures the hosts similarity from storage perspective, then I guess it could work. I think you know better if there is a risk that some user want to use loose labeling or does this new labeling logic even allow a bit heterogeneous nodes to have same label?

Reply:
> > Do you mean that in Machineset the device index/Name/HCTL number would be the same in all hosts that has label matching the MachineSet host requirements? In other words, after inspections, storage device list would be identical for all hosts having the label that match Machineset requirements
> 
> If labeling ensures the hosts similarity from storage perspective, then I guess it could work. I think you know better if there is a risk that some user want to use loose labeling or does this new labeling logic even allow a bit heterogeneous nodes to have same label?

There are 2 possible sources for the labels. 

1. Users will add their own to hosts manually. We have no control over that, so it would be up to the users to do it correctly, but I don't see that as a big problem. If users can manually set the device hint, they can manually set a label.
2. The automated host classification controller will add labels based on rules. In that case it's up to the user to define the rules correctly. That may be a little harder to get right, but if we design the API well it should end up being straightforward to do.

Reply:
> > > Do you mean that in Machineset the device index/Name/HCTL number would be the same in all hosts that has label matching the MachineSet host requirements? In other words, after inspections, storage device list would be identical for all hosts having the label that match Machineset requirements
> > 
> > 
> > If labeling ensures the hosts similarity from storage perspective, then I guess it could work. I think you know better if there is a risk that some user want to use loose labeling or does this new labeling logic even allow a bit heterogeneous nodes to have same label?
> 
> There are 2 possible sources for the labels.
> 
> 1. Users will add their own to hosts manually. We have no control over that, so it would be up to the users to do it correctly, but I don't see that as a big problem. If users can manually set the device hint, they can manually set a label.
> 2. The automated host classification controller will add labels based on rules. In that case it's up to the user to define the rules correctly. That may be a little harder to get right, but if we design the API well it should end up being straightforward to do.
In case 2. would it means that if user wants to use device name as a hint, then it would set such a labeling rule that each host that can be labeled with label X has to have device named Y(root device)

Reply:
> > There are 2 possible sources for the labels.
> > 
> > 1. Users will add their own to hosts manually. We have no control over that, so it would be up to the users to do it correctly, but I don't see that as a big problem. If users can manually set the device hint, they can manually set a label.
> > 2. The automated host classification controller will add labels based on rules. In that case it's up to the user to define the rules correctly. That may be a little harder to get right, but if we design the API well it should end up being straightforward to do.
> 
> In case 2. would it means that if user wants to use device name as a hint, then it would set such a labeling rule that each host that can be labeled with label X has to have device named Y(root device)

Yes, that's true. No matter which method of specifying a device root hint we use, we would use the label selector in the MachineSet to find hosts that should belong to that MachineSet.

Reply:
> > > There are 2 possible sources for the labels.
> > > 
> > > 1. Users will add their own to hosts manually. We have no control over that, so it would be up to the users to do it correctly, but I don't see that as a big problem. If users can manually set the device hint, they can manually set a label.
> > > 2. The automated host classification controller will add labels based on rules. In that case it's up to the user to define the rules correctly. That may be a little harder to get right, but if we design the API well it should end up being straightforward to do.
> > 
> > 
> > In case 2. would it means that if user wants to use device name as a hint, then it would set such a labeling rule that each host that can be labeled with label X has to have device named Y(root device)
> 
> Yes, that's true. No matter which method of specifying a device root hint we use, we would use the label selector in the MachineSet to find hosts that should belong to that MachineSet.
OK, if we can assume that hosts under the same Machineset has identical root device identity, could we expose set of hints in BMH spec, like subset of current spec: name, HCTL, size. So user would have a freedom in provisioning to decide which of the hint types to use?

Reply:
Could we take an intermediary approach ? Giving the WWN as hint being required, the solution could allow to specify something that is specific to the host such as the WWN while allowing for the Metal3Machine manager to select one of the disks based on some criteria and pass some info to identify it to BMO, that could be the WWN too. So if BMH exposes the root device hint and allows the user / higher layers to set it, then the end user could specify some requirements on the disk in the Metal3MachineTemplate (for the MachineSet), that the Metal3Machine manager would use to select one of the disks, get the WWN and pass it to BMO. This way we achieve both goals, with BMO behavior being fully predictable, and the choice left to the higher layers if needed.

Reply:
It was pointed out elsewhere that one reason for using WWN is for Ceph deployments. The inspection agent isn't going to necessarily see storage exposed from Ceph or other dynamic storage systems, especially if it isn't created and exposed already when the host CR is created and inspection is performed, since that may happen a long time before provisioning. We should make sure this case is explained in the document and give it as a reason why the hint cannot just be the index of into Status.HardwareDetails.Storage.

Reply:
It's not entirely clear what the conclusion is of this thread. Sorry if I'm just stating the obvious. It would help to get this thread incorporated into the proposal and marked as resolved, especially since I think the thread might be longer than the proposal itself. :wink: